Add support for exporting metrics to Prometheus.

zmagg opened the pull request:
### Description
WIP PR to export metrics to Prometheus, as discussed in https://github.com/vitessio/vitess/issues/3644

I've gotten feedback from @demmer on the approach in this PR, but I'd like to get some feedback from others as well, especially around metric names and the API exposed. Despite the long list of changes below, there are *no* breaking changes in this PR for anybody that relies on the expvar stats.
 
#### Metrics backend changes / naming stuff
* Adds `pull_backend.go` which supports registering generic pull backends which can scrape for metrics. 
* Adds a Prometheus backend that includes custom collectors per metric type and creates constant, throwaway, metrics "on the fly" at scrape time. The Prometheus backend is set up to be configured as a plugin per Vitess component. 
* The Prometheus backend also programmatically converts the camel case metric names to snake case metric names at scrape time, to align with the [Prometheus naming conventions](https://prometheus.io/docs/practices/naming/). This includes special casing names that don't convert easily (like VSchema) (See: [toSnake()](https://github.com/tinyspeck/vitess/blob/2ccc0533910921f101b48fb6fbedae8a2417802f/go/stats/prombackend/prombackend.go#L144)) . 
Right now, it also adds the name of the component as a namespace to the metric, but that produces some metric names that have duplicate prefixes, as we already prefix a lot of the Vitess metric names with the component type (_e.g. `VtgateVschemaCounts `turns into `vtgate_vtgate_vschema_count`_)

#### Instrumentation changes
* Adds Gauge-style types (Gauge, GaugeWithLabels, etc.) Prometheus expects Gauges and Counters to be separate types and we need this information in order to specify what kind of metric we're exporting.
* Renamed all metric types. Mapping of rename:
`Int` => `Counter`/`Gauge` (depending)
`IntFunc` => `GaugeFunc`
`Counters` => `CountersWithLabels` / `GaugesWithLabels`(depending)
`MultiCounters` => `CountersWithMultiLabels` /`GaugesWithMultiLabels` (depending)
`MultiCounterFunc` => `CountersFuncWithMultiLabels`

* Adds a description string per metric. 
* Adds an explicit labelName for the tags in `MultiCounters`. Prometheus's labels are all key/value pairs.
* Changes calls to `stats.Publish(<IntFunc>/<DurationFunc>)` to call `stats.NewIntFunc`/`stats.NewDurationFunc` explicitly instead.

#### TODO still
- [x] Add tests for each of the types of metrics
- [x] Add support for DurationFunc
- [x] Move registration of promhttp into servenv
- [x] After https://github.com/vitessio/vitess/pull/3816 is merged, modify it to support toSnakeCase and a few replacers instead.
- [x] Get TravisCI working with the new packages I've added to the vendor file (promhttp) 
- [x] Add a `prombackend` plugin for each component with metrics to be exported
- [x] It looks like many of the Counters/MultiCounters/MultiCounterFuncs are *actually Gauges* in practice. Since Prometheus differentiates between these two types, we need to go through and swap some of these out, where appropriate.
- [x] Dedupe the metric name with the namespace. e.g. _vtgate_vtgate_vschema_counts_ should just be _vtgate_vschema_counts_

#### Metric types not exported
- NewString / NewStringMap / PublishJSONFunc 
Prom. doesn't handle string values. Some of these could be converted into different types, if we'd like to export the metrics anyway. 

-  NewFloat 
This isn't called anywhere today. We could maybe export them anyway in case there are any callers in the future.

- direct calls to [stats.Publish(...CounterFunc()) ](https://github.com/vitessio/vitess/blob/master/go/vt/vttablet/tabletmanager/binlog_players.go#L434)
I think this should get ported in a followup PR, possibly refactoring it to use MultiCounterFunc.

#### Additional things to do in follow-on PRs
- Visit Rates / DurationFunc callsites and instrument them so that we can derive the rate/duration in Prometheus at querytime.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vitessio/vitess/3784)
<!-- Reviewable:end -->

sougou commented on go/stats/counters.go:
```diff
@@ -187,13 +229,61 @@ func (mc *MultiCounters) Add(names []string, value int64) {
+		publish(name, t)
+	}
+
+	publishPullMultiGauges(t, name)
```
I think you should be able to do this by registering a new var hook using `stats.Register`, and have that function make these calls. This way, there will be no need to spray these additional publish calls through all the types.

sougou commented on go/stats/export.go:
```diff
@@ -229,6 +231,53 @@ func (v *Int) String() string {
 }
 
+// Help returns the help string
+func (v *Int) Help() string {
```
For all these simple types `Int` is same as `IntGauge`. We might as well have only one of them. Maybe we can just rename `Int` to `IntGauge`.

sougou commented on go/stats/export.go:
```diff
@@ -263,11 +312,39 @@ func (v *Duration) String() string {
 // String is the implementation of expvar.var
-func (f IntFunc) String() string {
-	return strconv.FormatInt(f(), 10)
+func (intFunc IntFunc) String() string {
```
Go style recommends using single letter names for simple types like this. I don't necessarily agree with it, but it's better to follow the guidelines :).

sougou commented on go/stats/prombackend/prombackend.go:
```diff
@@ -0,0 +1,160 @@
+func Init(namespace string) {
+	http.Handle("/metrics", promhttp.Handler())
+	promBackend := &PromBackend{namespace: namespace}
+	stats.RegisterPullBackendImpl("prom", promBackend)
```
This should just call `stats.Register` with that new function that can call the various pull functions.

sougou commented on go/stats/prombackend/prombackend.go:
```diff
@@ -0,0 +1,160 @@
+	return output
+}
+
+func toSnake(name string) string {
```
@michael-berlin is there a way you can export youtube's function that converts to snake-case? I'm assuming that it shouldn't have anything proprietary.

zmagg replied:
Moving the snake case / kebab-case conversation here:

> (From @michael-berlin in Reviewable:) Turns out snake_case uses underscores but internally we're doing kebab-case (with hyphens).
I've exported the code: #3816
But I'm not sure if it's worth it merging it.
Also note that we do not have special handling for certain words e.g. the resulting internal variable name is vtgate-v-schema-counts :)

Ooh. Thanks for exporting! 

@sougou Thoughts here? I can also just change my `toSnakeCase` to follow what `toKebabCase` is doing with the `regexp` package (which looks significantly cleaner than my hack with runes), and add a test. 

I would also prefer to do some special casing (For things like VSchema, VtGate) as part of the Prometheus export.

demmer commented on go/stats/prombackend/collectors.go:
```diff
@@ -0,0 +1,214 @@
+	for i := range cutoffs {
+		key := float64(cutoffs[i]) / 1000000000
+		//TODO(zmagg): int64 => uint64 conversion. error if it overflows?
+		output[key] = uint64(buckets[i]) + last
```
It seems to me the buckets should just be uint64 from the start.

zmagg replied:
Hmm. They're `sync2.AtomicInt64`s in `histogram.go`.

demmer replied:
I think that's wrong too :)

Of course the "right" thing seems to me that we add a `sync2.AtomicUint64` and change all the histogram metrics to use that so we don't have to do this dance here.

@sougou any opinion here?

zmagg commented:
Updated description to account for new implementation changes and new TODOs. : )

demmer commented on go/stats/pull_backend.go:
```diff
@@ -0,0 +1,27 @@
+package stats
+
+// PullBackend should be implemented to export pull metrics from Vitess.
```
With the new change to just use the `stats.Publish` hook I actually don't think we need this interface any more since it looks like we never use it anywhere.

That also would let all the prom registration functions be private to the package instead of public.

demmer commented on go/stats/prombackend/collectors.go:
```diff
@@ -0,0 +1,214 @@
+
+// metricsCollector collects both stats.Counters and stats.Gauges
+type metricsCollector struct {
+	counters map[*stats.Counters]*prom.Desc
```
From what I can tell reading below, it looks like we never actually put more than one `stats.Counters` into any given metricsCollector, so can this be simplified to just have a pointer to the counter and a pointer to the `Desc`?

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -65,27 +67,48 @@ var (
 	)
 	// InternalErrors shows number of errors from internal components.
-	InternalErrors = stats.NewCounters("InternalErrors", "Task", "StrayTransactions", "Panic", "HungQuery", "Schema", "TwopcCommit", "TwopcResurrection", "WatchdogFail", "Messages")
+	InternalErrors = stats.NewCountersWithLabels("InternalErrors", "Number of errors from internal components", "component", "Task", "StrayTransactions", "Panic", "HungQuery", "Schema", "TwopcCommit", "TwopcResurrection", "WatchdogFail", "Messages")
```
I think a label key of "type" makes more sense.

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -65,27 +67,48 @@ var (
+	Warnings = stats.NewCountersWithLabels("Warnings", "Number of warnings", "type", "ResultsExceeded")
 	// Unresolved tracks unresolved items. For now it's just Prepares.
-	Unresolved = stats.NewCounters("Unresolved", "Prepares")
+	Unresolved = stats.NewGaugesWithLabels("Unresolved", "Tablet env's unresolved items", "item_type", "Prepares")
```
"Tablet env" is an internal abstraction. I'd just describe this as "Unresolved items"

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -65,27 +67,48 @@ var (
-	UserTableQueryCount = stats.NewMultiCounters("UserTableQueryCount", []string{"TableName", "CallerID", "Type"})
+	UserTableQueryCount = stats.NewCountersWithMultiLabels(
+		"UserTableQueryCount",
+		"Number of queries received for each CallerID/table comb",
```
"comb"=> "combination"

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -65,27 +67,48 @@ var (
-	UserTableQueryTimesNs = stats.NewMultiCounters("UserTableQueryTimesNs", []string{"TableName", "CallerID", "Type"})
+	UserTableQueryTimesNs = stats.NewCountersWithMultiLabels(
+		"UserTableQueryTimesNs",
+		"Shows total latency for each CallerID/table combo",
```
Again I think drop "Shows" and change "combo" to "combination"

demmer commented on go/vt/vttablet/tabletserver/tabletserver.go:
```diff
@@ -238,7 +238,7 @@ func NewTabletServer(config tabletenv.TabletConfig, topoServer *topo.Server, ali
 	// And we can remove the tsOnce variable.
 	tsOnce.Do(func() {
-		stats.Publish("TabletState", stats.IntFunc(func() int64 {
+		stats.NewGaugeFunc("TabletState", "Tablet server's state", (func() int64 {
```
"Tablet server state"

demmer commented on go/vt/vttablet/tabletserver/txserializer/tx_serializer.go:
```diff
@@ -40,21 +40,38 @@ var (
-	queueExceeded = stats.NewCounters("TxSerializerQueueExceeded")
+	queueExceeded = stats.NewCountersWithLabels(
+		"TxSerializerQueueExceeded",
+		"Number of transactions that were rejcted because the max queue size per row range was exceeded",
```
"Rejected"

demmer commented on go/vt/worker/worker.go:
```diff
@@ -59,52 +59,90 @@ var (
+		[]string{"Keyspace", "ShardName", "ThreadId"})
 	// statsStateDurations tracks for each state how much time was spent in it. Mainly used for testing.
-	statsStateDurationsNs = stats.NewCounters("WorkerStateDurations")
+	statsStateDurationsNs = stats.NewGaugesWithLabels("WorkerStateDurations", "How much time was spent in each state", "state")
```
Why not a Timer?

zmagg replied:
Hmm. It's a good idea. Followup PR? I'd like to avoid breaking more in this one. : )

michael-berlin commented:
Reviewed 5 of 43 files at r2, 3 of 19 files at r3, 3 of 12 files at r4.
Review status: 11 of 62 files reviewed at latest revision, 37 unresolved discussions.

---

*[go/cmd/vtgate/plugin_prombackend.go, line 27 at r4](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9XXl6c96Cucsp1MQDH:-L9XXl6c96Cucsp1MQDI:bgljyzz) ([raw file](https://github.com/vitessio/vitess/blob/667f2cede3b228b1f492666ccc964442d0836b03/go/cmd/vtgate/plugin_prombackend.go#L27)):*
> ```Go
> 
> func init() {
> 	prombackend.Init("vtgate")
> ```

Is it necessary to have a different namespace for each binary?

If so, what about other binaries which we have e.g. vtctld and vtworker?

Does Prometheus require that? Its internal Google counterpart doesn't care and instead can distinguish binaries by their production job name ;)

---

*[go/stats/export.go, line 266 at r4](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9dMQJN4sfqJPOA49tN:-L9XYX6H5s9nu1HaY9us:b-1rqyl7) ([raw file](https://github.com/vitessio/vitess/blob/c4a77123b9711233155e0508f1b2157c04ef32db/go/stats/export.go#L266)):*
> ```Go
> // IntFunc converts a function that returns
> // an int64 as an expvar.
> type IntFunc func() int64
> ```

As @demmer pointed out before, you'll probably need a `CounterFunc` which will be the renamed version of this.

Note that the removal/rename of this will break the Google internal plugins. Given that, please give me or @alainjobart a head's up before merging this. This way, we can change all internal callers first and verify that they won't break due to this change.

---

*[go/stats/prombackend/prombackend.go, line 144 at r1](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9XO6GeX47c1AhIPdGi-r1-144:-L9XXUJqDH1cIVvG0t8h:bdn5qoc) ([raw file](https://github.com/vitessio/vitess/blob/2ccc0533910921f101b48fb6fbedae8a2417802f/go/stats/prombackend/prombackend.go#L144)):*
<details><summary><i>Previously, sougou (Sugu Sougoumarane) wrote…</i></summary><blockquote>

@michael-berlin is there a way you can export youtube's function that converts to snake-case? I'm assuming that it shouldn't have anything proprietary.
</blockquote></details>

Turns out snake_case uses underscores but internally we're doing kebab-case (with hyphens).

I've exported the code: https://github.com/vitessio/vitess/pull/3816

But I'm not sure if it's worth it merging it.

Also note that we do not have special handling for certain words e.g. the resulting internal variable name is `vtgate-v-schema-counts` :)

---

*[go/stats/prombackend/prombackend.go, line 11 at r4](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9XWatm1IQZdHYVNvMM:-L9XWatm1IQZdHYVNvMN:b-en3wmq) ([raw file](https://github.com/vitessio/vitess/blob/667f2cede3b228b1f492666ccc964442d0836b03/go/stats/prombackend/prombackend.go#L11)):*
https://github.com/golang/go/wiki/CodeReviewComments#imports

> Avoid renaming imports except to avoid a name collision; good package names should not require renaming.

Given this style guide recommendation, you shouldn't rename the import to "prom".

(In general, I personally don't like abbreviations (unless it's a local variable or a Go receiver name). E.g. "prombackend" is less descriptive than "prometheus_backend". Similar comment for the other names e.g. "PublishPromMetric".)

---

*[go/vt/vtgate/masterbuffer/masterbuffer.go, line 50 at r4](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9dMA4l9MVcMZIJkoK5:-L9dMA4l9MVcMZIJkoK6:b-m9s6eb) ([raw file](https://github.com/vitessio/vitess/blob/667f2cede3b228b1f492666ccc964442d0836b03/go/vt/vtgate/masterbuffer/masterbuffer.go#L50)):*
> ```Go
> 	fakeBufferDelay        = flag.Duration("fake_buffer_delay", 1*time.Second, "The amount of time that we should delay all master requests for, to fake a buffer.")
> 
> 	bufferedRequestsAttempted  = stats.NewCounter("BufferedRequestsAttempted", "Count of buffered requests attempted")
> ```

Head's up: This is dead code which I'm going to delete here: https://github.com/vitessio/vitess/pull/3814

---

*[go/vt/worker/worker.go, line 63 at r4](https://beta.reviewable.io/reviews/vitessio/vitess/3784#-L9XPDYE2pavclaKkCSq:-L9XPDYE2pavclaKkCSr:b9wj01v) ([raw file](https://github.com/vitessio/vitess/blob/667f2cede3b228b1f492666ccc964442d0836b03/go/vt/worker/worker.go#L63)):*
> ```Go
> 	// statsRetryCount is the total number of times a query to vttablet had to be retried.
> 	statsRetryCount = stats.NewCounter("WorkerRetryCount", "Total number of times a query to a vttablet had to be retried")
> 	// statsRetryCount groups the number of retries by category e.g. "TimeoutError" or "Readonly".
> ```

Can you please delete the comments since you copied them into the description now?

Please make sure that no information gets lost e.g. the description for this stats var misses the examples `e.g. "TimeoutError" or "Readonly"`.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/vitessio/vitess/3784)*
<!-- Sent from Reviewable.io -->

zmagg commented:
@michael-berlin 

Replying here in a few comments, as I can't respond in Reviewable for now. (I need to check with some internal people about the permissions that responding in  Reviewable requires (it asked for write perms to *all* my orgs...?))

 Small stuff first:

> Is it necessary to have a different namespace for each binary?
If so, what about other binaries which we have e.g. vtctld and vtworker?
Does Prometheus require that? Its internal Google counterpart doesn't care and instead can distinguish binaries by their production job name ;)

@demmer and I talked about this a while back, but I forget why we landed on this. Prometheus does let you specify a namespace per binary in the static scrape config, which we could use instead of putting the namespace here. @demmer, any new thoughts?

 > Avoid renaming imports except to avoid a name collision; good package names should not require renaming.
Given this style guide recommendation, you shouldn't rename the import to "prom".
(In general, I personally don't like abbreviations (unless it's a local variable or a Go receiver name). E.g. "prombackend" is less descriptive than "prometheus_backend". Similar comment for the other names e.g. "PublishPromMetric".)

That sounds good actually, I'll do a rename pass. 

> Head's up: This is dead code which I'm going to delete here: #3814

Thanks for the head's up : )

> Can you please delete the comments since you copied them into the description now?
Please make sure that no information gets lost e.g. the description for this stats var misses the examples e.g. "TimeoutError" or "Readonly".

Will do!

zmagg commented on go/stats/export.go:
```diff
@@ -261,15 +228,6 @@ func (v *Duration) String() string {
 	return strconv.FormatInt(int64(v.i.Get()), 10)
 }
 
-// IntFunc converts a function that returns
```
> (From @michael-berlin in Reviewable): As @demmer pointed out before, you'll probably need a CounterFunc which will be the renamed version of this.
Note that the removal/rename of this will break the Google internal plugins. Given that, please give me or @alainjobart a head's up before merging this. This way, we can change all internal callers first and verify that they won't break due to this change.

Yes! It's changed to a `GaugeFunc` in https://github.com/vitessio/vitess/pull/3784/commits/a9d7fefbb58f6039dd0852c9531e1d6c3f3cc15f 
I think we'll also need a `CounterFunc`, for the ones that are Counters and not Gauges.

*re: breaking the Google internal plugins*

I'd love to understand how the Google internal plugins use this so that I can avoid breaking it. Are you using the expvar hook or some other way of exporting and converting the expvars to the internal system?

Is it the *removal* here that is likely to break (So, here, we went from `IntFunc` as a parameter to `stats.Publish` to explicitly creating a new `GaugeFunc`), or is it the any/all metric type renames? If it's the latter, head's up that I actually renamed almost all of the metric names. Some of these were to more clearly indicate the Gauge/Counter difference. (Full map in PR description). 

Will that require a lot of work on y'all's part to update the plugins internally?

michael-berlin replied:
Internal code basically just adds additional stats variables.

Examples:
```go
stats.Publish("XXX", stats.IntFunc(<some method>))

stats.Publish("XXX", stats.FloatFunc(func() float64 { return float64(<some method>) }))

timings:   stats.NewMultiTimings("VtgateXxxApi", []string{"Operation", "Keyspace", "DbType"}),

errorCounters = stats.NewCounters("XXXErrors", "A", "B")

requests = stats.NewMultiCounters("XXX", []string{"A", "B", "C"})
```

As long as it's easy to rewrite this code, it won't be a lot of work and fine.

zmagg commented:
FYI I pulled the conversation about removing `IntFunc` potentially breaking Google internal plugins to a comment here: https://github.com/vitessio/vitess/pull/3784/files#r180172907 so that we can have a thread of conversation about it. 

I pulled the conversation about `snakeCase` stuff back into Sougou's original comment about exporting the snake case functionality here:  https://github.com/vitessio/vitess/pull/3784#discussion_r178647700

michael-berlin commented on go/cmd/vtgate/plugin_prombackend.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func init() {
+	prombackend.Init("vtgate")
```
Starting a thread here instead of using Reviewable:

I wrote:

> Is it necessary to have a different namespace for each binary?
> If so, what about other binaries which we have e.g. vtctld and vtworker?
> Does Prometheus require that? Its internal Google counterpart doesn't care and instead can distinguish binaries by their production job name ;)

zmagg@ responded:

> @demmer and I talked about this a while back, but I forget why we landed on this. Prometheus does let you specify a namespace per binary in the static scrape config, which we could use instead of putting the namespace here. @demmer, any new thoughts?

The plugin here actually does two things:

1. Install the "/metrics" handler.
1. Set the namespace.

Would it make sense to move 1. to our servenv instead and install the handler by default? What are the downsides of having the handler in there by default?

re: 2. If setting the namespace is a crucial aspect, I would prefer to do it in code. This way it's more reusable. Otherwise, other people will ask you for this static map config ;) If you set the namespace in code, my comment is that you should set the namespace for all binaries, not just vtgate and vttablet.

zmagg replied:
Oh interesting. 

1. I think moving it into servenv makes sense, location-wise. The thought here was that we wanted to be able to not have any Prometheus-stuff running if a build didn't want it, but I think we can handle that inside servenv with a flag. 
2. Yeah, setting the namespace is crucial. I've added one for the `vtworker` and `vtctl` binaries now, are there any that I missed that metrics are in too? Those are all the binaries that we use at Slack.

zmagg commented on go/vt/vttablet/tabletserver/connpool/pool.go:
```diff
@@ -82,15 +82,15 @@ func New(
+	stats.NewGaugeFunc(name+"Active", "Tablet server conn pool active", cp.Active)
+	stats.NewGaugeFunc(name+"InUse", "Tablet server conn pool in use", cp.InUse)
+	stats.NewGaugeFunc(name+"MaxCap", "Tablet server conn pool max cap", cp.MaxCap)
+	stats.NewGaugeFunc(name+"WaitCount", "Tablet server conn pool wait count", cp.WaitCount)
```
@demmer similarly to comment above, did you mean the MaxCap one and not the WaitCount one?

demmer replied:
Same comment applies :)

zmagg commented on go/vt/dbconnpool/connection_pool.go:
```diff
@@ -65,15 +65,15 @@ func NewConnectionPool(name string, capacity int, idleTimeout time.Duration) *Co
+	stats.NewGaugeFunc(name+"Active", "Connection pool active", cp.Active)
+	stats.NewGaugeFunc(name+"InUse", "Connection pool in-use", cp.InUse)
+	stats.NewGaugeFunc(name+"MaxCap", "Connection pool max cap", cp.MaxCap)
+	stats.NewGaugeFunc(name+"WaitCount", "Connection pool wait count", cp.WaitCount)
```
@demmer Do you mean the max cap one? 

WaitCount seems like it must be a gauge: "WaitCount returns how many clients are waiting for a connection"

demmer replied:
Unfortunately that comment is inaccurate... WaitCount is in fact a counter if we peel away the onion and look at the underlying ResourcePool:

```
// WaitCount returns the total number of waits.
func (rp *ResourcePool) WaitCount() int64 {
	return rp.waitCount.Get()
}
```

This is only ever incremented.

zmagg commented:
@demmer Can you take another look? The PR description is up to date with my latest changes. : ) 

One thing I'm considering adding is preventing Counter-type metrics from calling `Add()` with negative numbers. Thoughts?

demmer commented on go/cmd/vtctld/plugin_prometheusbackend.go:
```diff
@@ -0,0 +1,35 @@
+
+func init() {
+	servenv.OnRun(func() {
+		http.Handle("/metrics", promhttp.Handler())
```
Rather than make each of the plugins pull in the promhttp details and register the handler, why not move this inside `prometheusbackend.Init`.

demmer commented on go/mysql/server.go:
```diff
@@ -42,10 +42,10 @@ const (
-	connSlow   = stats.NewInt("MysqlServerConnSlow")
+	timings    = stats.NewTimings("MysqlServerTimings", "MySQL server timings")
+	connCount  = stats.NewGauge("MysqlServerConnCount", "Active MySQL server connections")
+	connAccept = stats.NewCounter("MysqlServerConnAccepted", "Count of connections accepted by MySQL server")
```
Admittedly this is taste, but I find the use of "Count of" to be unnecessary in this and other descriptions. Ideally we should be consistent one way or the other and either always use it or never use it.

Personally I find "Accepted MySQL server connections" to be sufficiently descriptive and more consistent with the gauge case (where we don't say "Gauge of active MySQL server connections").

demmer commented on go/proc/counting_listener.go:
```diff
@@ -37,8 +38,8 @@ type countingConnection struct {
 		Listener:   l,
-		ConnCount:  stats.NewInt(countTag),
-		ConnAccept: stats.NewInt(acceptTag),
+		ConnCount:  stats.NewGauge(countTag, "Active connections accepted by counting listener"),
```
It has always been confusing to me that `ConnCount` and `ConnAccepted` don't have more descriptive names, and these generic descriptions don't help either :)

While I'm happy to defer this to a subsequent improvement, IMO we should change the `proc.Listen` API to include a description string and pass that here as well along with the countTag.

Then these would be:
```
ConnCount: stats.NewGauge(countTag, fmt.Sprintf("Active %s connections", description)),
ConnAccept: stats.NewCounter(countTag, fmt.Sprintf("Accepted %s connections", description)),
```

Since these are only used for HTTP, we would then change `servenv.Run` to include "HTTP" as the description string.

I also think we would be well served to rename the generic metric names as well to be `HTTPConnCount` and `HTTPConnAccepted` but that would break backwards compatibility. @sougou / @michael-berlin thoughts on this?

zmagg replied:
Agreed, but let's do it in a follow-on?

demmer replied:
Yup. Would still like to hear from @sougou and @michael-berlin about the backwards-incompatibility of this proposed change.

demmer commented on go/stats/counters.go:
```diff
@@ -19,36 +19,85 @@ package stats
+}
+
+// Add adds the provided value to the Counter
+func (v *Counter) Add(delta int64) {
```
FWIW the prom API includes an `Inc()` method which seems nicer than `Add(1)`, especially if we follow through with the idea to prevent Add of a negative number.

zmagg replied:
Added logging for now!

demmer commented on go/stats/prometheusbackend/collectors.go:
```diff
@@ -0,0 +1,204 @@
+	ch <- prometheus.MustNewConstMetric(c.desc, prometheus.CounterValue, float64(c.df.F()/1000000000.0))
+}
+
+func toPromValueType(vt ValueType) prometheus.ValueType {
```
Why do we need this conversion and the `ValueType` enum at all? It seems like holdover from the more generic pullbackend design, and with this current approach we can avoid the conversion by just passing the appropriate prometheus metric type along when we instantiate the collector.

demmer commented on go/vt/dbconnpool/connection_pool.go:
```diff
@@ -65,15 +65,15 @@ func NewConnectionPool(name string, capacity int, idleTimeout time.Duration) *Co
+	stats.NewGaugeFunc(name+"MaxCap", "Connection pool max cap", cp.MaxCap)
+	stats.NewGaugeFunc(name+"WaitCount", "Connection pool wait count", cp.WaitCount)
+	stats.NewDurationFunc(name+"WaitTime", "Connection pool wait time", cp.WaitTime)
+	stats.NewDurationFunc(name+"IdleTimeout", "Connection pool idle timeout", cp.IdleTimeout)
```
Ugh... looking at this deeper I'm afraid to say that we need the ability for a duration to map to either a prom counter or a prom gauge.

In these two examples, WaitTime is a counter since (like WaitCount) it is only ever incremented, while IdleTimeout is a gauge since it represents the configured timeout.

To that end we have a couple options... the simplest would be to have `NewDurationGaugeFunc` and `NewDurationCounterFunc` which bloats the interface but at the same time is most consistent with the other cases.

Another option would be to overload the existing `stats.NewGaugeFunc` and `stats.NewCounterFunc` interfaces with an optional conversion function which would implement the mapping, much like how the old API was.

This way the caller API would be:
`stats.NewCounterFunc(name+"WaitTime", "Connection pool wait time", stats.DurationFunc(cp.WaitTime))`

Then just like before, the implementation of `stats.DurationFunc` would allow either expvar to get the String version or prom to get the underlying Duration.

FWIW I _think_ that would allow the prom implementation to call `duration.Seconds` instead of going to nanoseconds and dividing which would reduce rounding errors.

zmagg replied:
re: WaitCount 

ohhhhh! I didn't peel the next layer of the onion. Wow! 

``` 
func (rp *ResourcePool) recordWait(start time.Time) {
        rp.waitCount.Add(1)
        rp.waitTime.Add(time.Now().Sub(start))
}
```

It only ever increments! Also, this seems like WaitCount & WaitTime should be condensed into a histogram/time.

zmagg replied:
But I think for now keeping this separated and making DurationFunc support both counters and gauges sounds good.

demmer commented on go/vt/srvtopo/resilient_server.go:
```diff
@@ -215,7 +215,7 @@ func NewResilientServer(base *topo.Server, counterPrefix string) *ResilientServe
 		cacheTTL:     *srvTopoCacheTTL,
 		cacheRefresh: *srvTopoCacheRefresh,
-		counts:       stats.NewCounters(counterPrefix + "Counts"),
+		counts:       stats.NewCountersWithLabels(counterPrefix+"Counts", "resilient srvtopo server operations", "type"),
```
Another naming nitpick... descriptions should either all start with a capital letter or not :)

demmer commented on go/vt/vtgate/buffer/variables.go:
```diff
@@ -23,22 +23,34 @@ import "vitess.io/vitess/go/stats"
-	starts = stats.NewMultiCounters("BufferStarts", []string{"Keyspace", "ShardName"})
+	starts = stats.NewCountersWithMultiLabels(
+		"BufferStarts",
+		"How many times we started buffering, including dry-run",
```
Again with the descriptions... this is the only place where we use "How many times we..."
I think the more impersonal:

"Buffering operation starts"
"Buffering operation stops"
"Total buffering failover duration"
"Cumulative buffer utilization % during failover"
"Cumulative buffer utilization % during failover (dry-run)"
"Buffered requests"
"Buffered requests (dry-run)"
"Completed buffering requests"
"Evicted buffering requests"
"Skipped buffering requests"

demmer commented on go/vt/vtgate/executor.go:
```diff
@@ -104,10 +104,10 @@ func NewExecutor(ctx context.Context, serv srvtopo.Server, cell, statsName strin
+		stats.NewGaugeFunc("QueryPlanCacheLength", "Query plan cache length", e.plans.Length)
+		stats.NewGaugeFunc("QueryPlanCacheSize", "Query plan cache size", e.plans.Size)
+		stats.NewGaugeFunc("QueryPlanCacheCapacity", "Query plan cache capacity", e.plans.Capacity)
+		stats.NewGaugeFunc("QueryPlanCacheEvictions", "Query plan cache evictions", e.plans.Evictions)
```
QueryPlanCacheEvictions is a CounterFunc

demmer commented on go/vt/vtgate/executor.go:
```diff
@@ -104,10 +104,10 @@ func NewExecutor(ctx context.Context, serv srvtopo.Server, cell, statsName strin
+		stats.NewGaugeFunc("QueryPlanCacheSize", "Query plan cache size", e.plans.Size)
+		stats.NewGaugeFunc("QueryPlanCacheCapacity", "Query plan cache capacity", e.plans.Capacity)
+		stats.NewGaugeFunc("QueryPlanCacheEvictions", "Query plan cache evictions", e.plans.Evictions)
 		stats.Publish("QueryPlanCacheOldest", stats.StringFunc(func() string {
```
Again this can be deferred to a future PR, but this seems like a case in which the earlier proposal to add a conversion API could be helpful -- that way we could export this as a gauge.

demmer commented on go/vt/vttablet/tabletserver/query_engine.go:
```diff
@@ -234,24 +234,24 @@ func NewQueryEngine(checker connpool.MySQLChecker, se *schema.Engine, config tab
+		stats.NewGaugeFunc("WarnResultSize", "Query engine warn result size", qe.warnResultSize.Get)
+		stats.NewGaugeFunc("MaxDMLRows", "Query engine max DML rows", qe.maxDMLRows.Get)
+		stats.NewGaugeFunc("StreamBufferSize", "Query engine stream buffer size", qe.streamBufferSize.Get)
+		stats.NewGaugeFunc("TableACLExemptCount", "Query engine table ACL exempt count", qe.tableaclExemptCount.Get)
```
TableACLExemptCount is a Counter.

demmer commented on go/vt/vttablet/tabletserver/query_engine.go:
```diff
@@ -234,24 +234,24 @@ func NewQueryEngine(checker connpool.MySQLChecker, se *schema.Engine, config tab
+		stats.NewGaugeFunc("QueryCacheLength", "Query engine query cache length", qe.plans.Length)
+		stats.NewGaugeFunc("QueryCacheSize", "Query engine query cache size", qe.plans.Size)
+		stats.NewGaugeFunc("QueryCacheCapacity", "Query engine query cache capacity", qe.plans.Capacity)
+		stats.NewGaugeFunc("QueryCacheEvictions", "Query engine query cache evictions", qe.plans.Evictions)
```
QueryCacheEvictions is a counter

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -34,18 +34,20 @@ import (
 	QPSRates = stats.NewRates("QPS", QueryStats, 15*60/5, 5*time.Second)
 	// WaitStats shows the time histogram for wait operations
-	WaitStats = stats.NewTimings("Waits")
+	WaitStats = stats.NewTimings("Waits", "Shows the time histogram for wait operations")
```
More description nits: "Shows the..." should be just "Wait operations" I think

demmer commented on go/vt/vttablet/tabletserver/tabletenv/tabletenv.go:
```diff
@@ -34,18 +34,20 @@ import (
+	// ErrorStats shows number of critial errors happened.
+	ErrorStats = stats.NewCountersWithLabels(
 		"Errors",
+		"Number of critical errors that happened",
```
I think:

"Critical errors"
"Internal component errors"
...

zmagg commented:
Ok! 

Responded to your comments. 

Major changes:
- Changed up `CounterFunc` & `GaugeFunc` to take in a new type: `MetricFunc` that allows us to both export ints (like it did before) and now also durations. 

I did a best effort pass through of which of the DurationFuncs were gauges/counters, but would love an extra eye on that @demmer. 

Also rebased, made all the other changes I 👍'd to your comments on.

What's left:
- Now that the toKebabCase 🍢 is merged, I'll take a look next week when I'm back to make the toSnake modifications we talked about in person. 
- As discussed in person, add logging / increment a counter if we ever call `Add()` with a negative number on a Counter. In future PRs, we can maybe ban such things once we're confident we've removed all existing cases from the codebase.

zmagg commented on go/stats/kebab_case_converter.go:
```diff
@@ -22,6 +22,32 @@ import (
+// For systems (like Prometheus) that ask for snake-case names.
+// It converts CamelCase to camel_case, and CAMEL_CASE to camel_case.
+// For numbers, it converts 0.5 to v0_5.
+func toSnakeCase(name string) (hyphenated string) {
```
I added the toSnakeCase modifications based on the toKebabCase stuff.  I'm not totally convinced it structurally makes sense to put it here (at minimum, I should rename the file if it otherwise makes sense to you). 

Also, it shares a memoizer with the toKebabCase, which _should_ be fine (one build should only use one or the other), but perhaps they should be separate in the code too? 

@demmer, thoughts?

demmer replied:
For safety and cleanliness I actually don't think they should share a memoizer.

Given that... I don't think there's actually _any_ substantive shared code any more, so rather than misleadingly sharing the file name, I'd say you should just create a `snake_case_converter.go` with the new code and a separate memoizer.

zmagg commented:
@demmer 
ok, finished up the rest of the items, although I have a few questions about where to put the `toSnakeCase` stuff structurally (commented in the code here: https://github.com/vitessio/vitess/pull/3784/commits/e2982789d53dafa882ca8c39ba7a57270fa823ab#r182589219 )

Thoughts now?

demmer commented:
From my standpoint the only remaining item before I think this is good to merge is the minor reorganization of where the snake conversion goes.

@michael-berlin / @alainjobart since this will entail some breaking changes with internal Google importing I would like a 👍 from one of you that this is ok to merge.

@sougou told me in DM he doesn't care, so he doesn't get a vote on this one.

sougou commented:
I do care, but I know it's in good hands with @michael-berlin and @alainjobart :)

michael-berlin commented on go/stats/counters.go:
```diff
@@ -19,36 +19,95 @@ package stats
+// Add adds the provided value to the Counter
+func (v *Counter) Add(delta int64) {
+	if delta < 0 {
+		log.Warningf("Adding a negative value to a counter, %v should be a gauge instead", v)
```
How likely is it that this will get called?

If it's not likely at all because you already checked all callers in open-source, then I'm okay with it as is.

Otherwise, it may be a good idea to use a throttled logger here instead. This way we would avoid that bad code spams the log, makes the disk full and potentially slows down the process.

zmagg replied:
Oh, that's a good idea. I've checked all the easy sites (`Add(-1)`) but there's a lot more that wind their way through the code that I don't have confidence on, hence logging for now instead of banning decrementing all together. I switched to using a throttled logger here. I also switched to using a throttled logger for the unsupported export type logger in `prometheusbackend.go`, thanks for the suggestion!

michael-berlin commented:
It looks like this is good to merge? If so, let's do as follows:

1. I'll try to import it locally and make all necessary changes.
2. If thinks are working, I'll submit the internal import.
3. Shortly after that, I'll merge this PR.

Let me try to get to this today or tomorrow.

zmagg commented:
@michael-berlin That sounds awesome! 

Let me figure out what's going on with these failing TravisCI tests after my last commit. The tests still succeed for me locally, so not sure. Debugging now.

michael-berlin commented on go/stats/counters.go:
```diff
@@ -19,36 +19,99 @@ package stats
+	"time"
+
+	"vitess.io/vitess/go/vt/logutil"
+
```
nit: Please remove the newline here.

michael-berlin commented on go/stats/counters.go:
```diff
@@ -19,36 +19,99 @@ package stats
-// it doesn't allow floats. In addition, it provides
-// a Counts method which can be used for tracking rates.
+// it doesn't allow floats. It is used to build CountersWithLabels and GaugesWithLabels.
 type Counters struct {
```
Is it actually necessary to export this?

I saw that you're using it in the promstatsbackend but the usage was not immediately clear to me.

I'm asking because we have several type switches in the code base and I made the mistake of using "stats.Counter" instead of "stats.CountersWithLabels" in these type switches.

zmagg replied:
Yeah I agree, it doesn't make sense. Let me subsume what it's doing into CountersWithLabels & GaugesWithLabels.

michael-berlin commented on go/stats/counters.go:
```diff
@@ -122,6 +185,159 @@ func (c *Counters) Counts() map[string]int64 {
+
+// CountersWithLabels provides a labelName for the tagged values in Counters
+// It provides a Counts method which can be used for tracking rates.
+type CountersWithLabels struct {
```
Technically, this is only a single label and not multiple labels.

i.e. it should be named `CountersWithLabel` ?

michael-berlin replied:
Okay. Let's keep it as is for now.

michael-berlin commented on go/stats/export_test.go:
```diff
@@ -71,38 +71,49 @@ func TestFloat(t *testing.T) {
+}
+
+func TestGaugeFunc(t *testing.T) {
+	var gotname string
```
These two variables are unused.

Same for `TestDurationFunc` below.

Also: The `Register() ` here seems to be unnecessary as well?

Can you please double-check this file and report back? Thanks!

michael-berlin commented:
I have an internal import open with all the necessary changes to our internal code. Tests are passing now.

Can you please look at my last comments and address them in additional CLs? Once that's done, I'll patch them into my pending import as well and then we can wrap things up.

zmagg commented:
@michael-berlin Yay, awesome! I responded to your comments and rebased to resolve the merge conflict.

michael-berlin commented:
I ran into the merge conflict during the import as well. Therefore, I just reverted that PR.

Given that, can you please rebase again against the latest master?

michael-berlin commented:
I just did it myself and pushed it to your branch.

michael-berlin commented:
Internal import is done.

I'll merge this as soon as Travis has passed.

michael-berlin commented:
The race unit test did flake and therefore test shard 0 failed. I'm ignoring that and will merge this now.
